File: packages/coding-agent/src/utils/deprecation.ts
import { t } from "@earendil-works/pi-tui";
import chalk from "chalk";

const emittedDeprecationWarnings = new Set<string>();

export function warnDeprecation(message: string): void {
	if (emittedDeprecationWarnings.has(message)) return;
	emittedDeprecationWarnings.add(message);
	console.warn(chalk.yellow(t("codingAgent.packageManager.success.deprecatedWarning", { message })));
}

/** Clear deprecation warning state. Exported for tests. */
export function clearDeprecationWarningsForTests(): void {
	emittedDeprecationWarnings.clear();
}

File: python/src/pi_mono/utils/deprecation.py
import sys

emitted_deprecation_warnings: set[str] = set()


def warn_deprecation(message: str) -> None:
    """Emit a deprecation warning in yellow to stderr once per unique message."""
    if message in emitted_deprecation_warnings:
        return
    emitted_deprecation_warnings.add(message)
    sys.stderr.write(f"\033[33mDeprecation warning: {message}\033[0m\n")
    sys.stderr.flush()


def clear_deprecation_warnings_for_tests() -> None:
    """Clear deprecation warning state. Exported for tests."""
    emitted_deprecation_warnings.clear()
